### ISSUE TITLE
apps and user names

### ISSUE THREAD

**Issue by kitenetter** [NERC-CEH/irecord-app#203]
This is a generic request across all our apps, taking the iRecord app as an example.

Indicia uses three types of person identifier for most records:
1. A username associated with the registered account
2. An email address associated with the registered account
3. A recorder name associated with each individual sample, but often pre-populated from the registered account

If a user registers via the iRecord app, their email address is used as their username. This can lead to breaches of our Privacy Notice, as usernames are made public in some situations (e.g. where a user adds a comment to an existing record).

I think we need to stop using the email address as the default username, and instead allow users to add their own choice of username at the point of registration.

See also #189 and #190 regarding making our use of recorder names more consistent.

**Comment by kazlauskis** [comment]
Thanks, yeah we shouldn't be leaking emails to the public. Since we also ask for a name and surname it should be those used instead of the username or the account email. So I would say it is better to think of how to autogenerate and hide a random username rather than ask it to enter it for the user (in which case usernames even as emails would be hidden).

**Comment by kitenetter** [comment]
I think there is value in allowing a user name, as well as a recorder name. Although we encourage people to use their real name when they are the recorder of a biological record, there is no need to do so for other purposes such as adding comments to records or to forums etc., and some people prefer not to use real names in those circumstances.

So I would argue that we need:
- email address (which is never publicly displayed)
- username (which is displayed)
- recorder name (which is shown with the record details but may be hidden in reports)

So I think we need to give users the chance to enter a user name that is different to their recorder name. We could pre-populate that user name with the recorder name (or perhaps with a random name), and allow the user to over-write it they wish.

**Comment by kazlauskis** [comment]
I don't know, it sounds to me that the first name and last name should be the 'display' name for the account. These are mandatory so if people are OK to have that shown publicly as part of their records on the Explore page I would think that the same should be shown for the record comments or forum discussions. It's your call but I suggest to simplify the current system to emails and real names.

**Comment by kitenetter** [comment]
If we follow your simplified suggestion, does that mean that the apps will continue to use "forename surname", which conflicts with the website "surname, forename"?

**Comment by kazlauskis** [comment]
Yes, we currently don't capture the `recorders` field in the **single record mode** in the app and the field is prepopulated in the backend so this wouldn't change. Since we will be adding the `recorders` field in the app (#189) we can make sure it is using `surname, forename` format.  

For the **survey modes** in the app we do capture the recorders but in `forename surname` format and I can correct it so it is consistent with the backend format (#190).

**Comment by kitenetter** [comment]
@kazlauskis we have discussed this further and have come to the conclusion that we should standardise as far as possible on `forename surname`, and NOT `surname, forename`.

So the current version of app has the correct format (and it is the website that we need to change).

I will close #190

This issue needs to remain open to ensure that email addresses are not stored as user names.

**Comment by kazlauskis** [comment]
So that I understand this correctly, would you like to add a `username` option to the registration or will the website be modified to not show the usernames?

**Comment by kitenetter** [comment]
@kazlauskis I think we need to retain a `username` as part of the registration, for compatibility with other parts of Indicia.

So the question is whether we ask the user to enter a username, or whether we just auto-populate it with the actual name that they give.

My preferred option would be for the registration process to be:
1. Enter an `email address`
2. Enter `first name and surname`
3. Ask for a username that is populated by default with `first name and surname` but allows the user to over-write this if they wish

That is similar to the settings used in iNaturalist:

![image](https://user-images.githubusercontent.com/16039111/110337469-dcb21f80-801d-11eb-9809-82b67b681cbb.png)

However, if that is too complex we could just auto-populate the username with `first name and surname` and not give the user a choice.

@DavidRoy can you see any arguments for or against enabling a user-defined username?

**Comment by japonicus** [comment]
'First name Surname' is very far from unique and although automated ways of making unique usernames would be possible (e.g. numerical suffixes), those are ugly - it's probably necessary to allow a user to choose their own.

Perhaps more importantly, there might be situations where someone is happy for their username to be exposed, but not their real name - making the two equivalent removes that choice. For that reason I think it might be best for people to have free choice of username, without an initial suggestion.

There's probably need for validation that rejects attempts to put an email address in the username field.

Does the app allow participation by anonymous users? I know there are strong arguments why occurrence records should have real names attached, but I'm uncertain if that should be mandatory. A partial solution might be to allow people to register with only a username.

**Comment by kazlauskis** [comment]
Thanks, yeah, the username must be unique so the autogenerated one should be available to edit.

**Comment by kitenetter** [comment]
I've had to revisit the name format issue today, and as far as I can tell app records are currently being displayed within the iRecord website with a recorder name formatted as "surname, forename". Which means it is consistent with the website itself.

I still think there is an argument to be made for changing both app and website to "forename surname", but I'm not proposing to revisit that at the moment, and am happy that things are consistent. Closing this issue.
